### PR TITLE
[Dynamic RPC] optional rank argument in init_rpc

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#77560 [Dynamic RPC] optional rank argument in init_rpc**

Make rank argument also optional when world_size is not passed in (dynamic rpc). Under the hood, the rank is still allocated from the store because it needs to be used for autograd and profiler, in the future this could be improved. This internal rank increments whenever a new worker joins the group.

New initialization looks like:

##### Filestore
```python
rbo = rpc.RpcBackendOptions(init_method="file://tempfile")
rpc.init_rpc("workerN", rpc_backend_options=rbo)
```

##### TcpStore
```python
# export MASTER_ADDR=localhost
# export MASTER_PORT=5678
# Process 1: master store is on rank 0 and must be specified to use TCPStore
rpc.init_rpc("main", rank=0)
# Process n: Rank of workers will be allocated
rpc.init_rpc("workerN")
```

### CODE REVIEW

**Review by mrshenli — COMMENTED** [review]
Would I be correct that `rank=0` is only needed for TCPStore but not FileStore?

File: torch/distributed/rpc/__init__.py [hunk 1]
```diff
@@ -165,6 +166,10 @@ def init_rpc(
             # If world_size is not set in construction and also not set in environment variables
```
Comment (mrshenli) [inline torch/distributed/rpc/__init__.py:166]: nit: this is prior to this PR. In the comment above:

"Create store, performs rendezvous" -> "Create store, perform rendezvous"?

**Comment by H-Huang** [comment]
> Would I be correct that rank=0 is only needed for TCPStore but not FileStore?

@mrshenli yes, that's correct

File: torch/distributed/rpc/__init__.py [hunk 2]
```diff
@@ -165,6 +166,10 @@ def init_rpc(
...
             # The store will be created for the dynamic group setting
             store = dist._create_store_from_options(rpc_backend_options, rank)
+            # If there is unspecified rank or rank is pre-allocated 0 (master), then rank will be allocated by store
+            if rank <= 0:
+                val = store.add(STORE_GET_RANK_KEY, 1)
+                rank = val - 1
```
Comment (mrshenli) [inline torch/distributed/rpc/__init__.py:172]: for rank > 0, what happens if it is duplicated with an auto-allocated rank? Say I do the following

init_rpc("A")
int_rpc("B")
int_rpc("C", rank=1)

File: torch/distributed/rpc/__init__.py [hunk 3]
```diff
@@ -165,6 +166,10 @@ def init_rpc(
             # If world_size is not set in construction and also not set in environment variables
             # The store will be created for the dynamic group setting
             store = dist._create_store_from_options(rpc_backend_options, rank)
+            # If there is unspecified rank or rank is pre-allocated 0 (master), then rank will be allocated by store
+            if rank <= 0:
+                val = store.add(STORE_GET_RANK_KEY, 1)
```
Comment (mrshenli) [inline torch/distributed/rpc/__init__.py:171]: what does `store.add` return on the first call? 0 or 1?

**Pull request closed without merging**